### PR TITLE
Fixed duplicate user/customer data in SSO/alfa after associating existing subscription or upgrading trial subscription from dummy address.

### PR DESCRIPTION
**Pull request by sun**
The client now needs to pass the email address to the checkAboNo API to allow the SSO to additionally check whether the subscriber ID is associated with a different account already.

And if the checkAboNo validation was correct, the client needs to pass the validated subscriber ID to registerUserAndPurchase and registerPurchase APIs, so that the SSO does not invoke registerUser towards alfa, but essentially does the opposite and imports the data from that subscriber ID from alfa into the SSO user account.

The same procedure also applies to the updateUser API when a user attempts to associate their existing subscription to their user account.

Lastly, a user should still be able to associate their existing subscription as long as there is only a "dummy address" in alfa, which refers to a partial address for trial subscriptions whose otherwise required fields have been filled up with placeholder values.  ssoFACT announces these addresses as "alfa_dummy_address" in the UserInfo now.

In short:  If it's not a alfa_dummy_address, the user is always able to associate.

- https://jira.newsfactory.de/browse/HS-115
- https://jira.newsfactory.de/browse/HS-126
- https://app.asana.com/0/29283351247093/772388816921913
- https://app.asana.com/0/29283351247093/784199051178313
- https://app.asana.com/0/29283351247093/784199051178314

### CODE REVIEW

File: src/Plugin.php [hunk 1]
```diff
@@ -642,16 +643,18 @@ public static function buildUserInfo($key_prefix = 'billing', $user_id = NULL) {
...
-      $remove_address = $user_id > 0 && (!empty($address_source['billing_subscriber_id']) || get_user_meta($user_id, 'billing_subscriber_id', TRUE));
-    }
-    else {
-      $remove_address = $user_id > 0 && get_user_meta($user_id, 'billing_subscriber_id', TRUE);
-    }
+    // reject any kind of change to the address, unless it is a "dummy address";
+    // i.e., a partial address record only consisting of salutation, firstname,
+    // lastname, used for quicker trial subscription signups, which is filled up
+    // by alfa with placeholder values for all other required address fields.
+    // If addresss values are removed, the submitted address will only be
```
Comment (sun) [inline src/Plugin.php]: 1. Typo in "addresss"

1. _"If address values are removed **below**, …"_ might be more clear

File: src/Plugin.php [hunk 2]
```diff
@@ -642,16 +643,18 @@ public static function buildUserInfo($key_prefix = 'billing', $user_id = NULL) {
...
+    // contained in the order notification email to the shop admin and manually
+    // processed by the customer service team.
+    // When processing a checkout, additionally check the checkAboNo data.
+    $remove_address = $user_id > 0 && !get_user_meta($user_id, 'alfa_dummy_address', TRUE);
+    // if (!$remove_address && is_checkout() && ($subscriber = WC()->session->get('subscriber_data'))) {
+    //   $userinfo['subscriber_id'] = $subscriber['subscriber_id'];
+    // }
```
Comment (sun) [inline src/Plugin.php]: Remove obsolete code

File: src/WooCommerce.php [hunk 3]
```diff
@@ -554,15 +559,15 @@ public static function woocommerce_billing_fields($fields) {
...
-    elseif (($subscriber = WC()->session->get('subscriber_data')) || !empty($subscriber_id)) {
-      if (empty($subscriber) && !empty($subscriber_id)) {
+    elseif (($subscriber = WC()->session->get('subscriber_data')) || !$subscriber_association_allowed) {
+      if (empty($subscriber)) {
         $subscriber = [
-          'abono' => $subscriber_id,
-          'salutation' => get_user_meta(get_current_user_ID(), 'billing_salutation', TRUE),
-          'lastname' => get_user_meta(get_current_user_ID(), 'billing_last_name', TRUE),
-          'company_contact' => get_user_meta(get_current_user_ID(), 'billing_company_contact', TRUE),
+          'subscriber_id' => !empty($subscriber_id) ? $subscriber_id : '???',
```
Comment (sun) [inline src/WooCommerce.php]: Case of "???" is not handled yet

**Comment by sun** [comment]
This PR is ready for final reviews / approval.
